### PR TITLE
5719: Add SMTP health check

### PR DESCRIPTION
**Pull request by assoomro**

### CODE REVIEW

File: src/NHSD.BuyingCatalogue.Identity.Common/Constants/HealthCheckTags.cs [hunk 1]
```diff
@@ -0,0 +1,8 @@
+﻿namespace NHSD.BuyingCatalogue.Identity.Common.Constants
+{
+    public static class HealthCheckTags
+    {
+        public const string Live = "live";
+        public const string Ready = "ready";
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Identity.Common/Constants/HealthCheckTags.cs:6]: `internal` would be sufficient for the class and constants here.

File: src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs [hunk 2]
```diff
@@ -69,6 +73,23 @@ public void ConfigureServices(IServiceCollection services)
                     }
                 });
 
+            services.AddHealthChecks()
+                .AddCheck(
+                    name: "self",
+                    check: () => HealthCheckResult.Healthy(),
+                    tags: new[] { HealthCheckTags.Live })
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs:80]: I'm not sure specifying the argument names is adding much here. Personally, I prefer it without them but see what you think.

File: src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs [hunk 3]
```diff
@@ -69,6 +73,23 @@ public void ConfigureServices(IServiceCollection services)
...
+                .AddSmtpHealthCheck(
+                    setup: (smtp) =>
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs:82]: The brackets for the lambda parameter can be omitted here.

File: src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs [hunk 4]
```diff
@@ -69,6 +73,23 @@ public void ConfigureServices(IServiceCollection services)
...
+                    {
+                        smtp.Host = smtpSettings.Host;
+                        smtp.Port = smtpSettings.Port;
+                        smtp.ConnectionType = SmtpConnectionType.TLS;
+                    },
+                    name: "smtp",
+                    failureStatus: HealthStatus.Unhealthy,
+                    tags: new[] { HealthCheckTags.Ready },
+                    timeout: TimeSpan.FromSeconds(10));
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs:91]: As above, I'm not sure specifying the argument names is adding much here. Personally, I prefer it without them but see what you think.

File: src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs [hunk 5]
```diff
@@ -104,6 +125,15 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env, ILogger<
             app.UseEndpoints(endpoints =>
             {
                 endpoints.MapControllers();
+                endpoints.MapHealthChecks("/health/live", new HealthCheckOptions
+                {
+                    Predicate = (healthCheckRegistration) => healthCheckRegistration.Tags.Contains(HealthCheckTags.Live)
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs:130]: The brackets can be omitted from the lambda parameter here.

File: src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs [hunk 6]
```diff
@@ -104,6 +125,15 @@ public void Configure(IApplicationBuilder app, IWebHostEnvironment env, ILogger<
...
+                });
+
+                endpoints.MapHealthChecks("/health/ready", new HealthCheckOptions
+                {
+                    Predicate = (healthCheckRegistration) => healthCheckRegistration.Tags.Contains(HealthCheckTags.Ready)
```
Comment (bjss-chris) [inline src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs:135]: As above, the brackets can be omitted from the lambda parameter here.

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Healthcheck.3538.feature [hunk 7]
```diff
@@ -0,0 +1,16 @@
+﻿Feature: Buying Catalogue Organisations healthchecks
+	As BuyingCatalogueOrganisations
+	I want to be be able to check the health of my dependencies
+    So that I can behave accordingly
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Healthcheck.3538.feature:4]: Spacing/alignment – it looks like there's a mix of tabs and spaces here. @CallumBradburyBJSS had a similar issue. I can't remember whether we want all tabs or all spaces – @CallumBradburyBJSS can you confirm which it should be?

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Healthcheck.3538.feature [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+Scenario: 1. Smtp Server is up
+	Given The Smtp Server is up
+	When the dependency health-check endpoint is hit
+	Then the response will be Healthy
+
+@3538
+Scenario: 2. Smtp Server is down
+	Given The Smtp Server is down
+	When the dependency health-check endpoint is hit
+	Then the response will be Unhealthy
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Healthcheck.3538.feature:16]: Finish with a newline (for git).

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Utils/Settings.cs [hunk 9]
```diff
@@ -10,11 +10,14 @@ public sealed class Settings
 
         public string OrganisationApiBaseUrl { get; }
 
+        public string BrokenOrganisationApiBaseUrl { get; }
+
         public Settings(IConfigurationRoot config)
```
Comment (bjss-chris) [inline tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Utils/Settings.cs:15]: The type can be `IConfiguration` here (as we're not currently using either of the additional couple of things that `IConfigurationRoot` brings in.

File: src/NHSD.BuyingCatalogue.Identity.Common/Constants/HealthCheckTags.cs (hunk 1 above)
Comment (assoomro) [reply src/NHSD.BuyingCatalogue.Identity.Common/Constants/HealthCheckTags.cs:6]: 👍

File: src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs (hunk 2 above)
Comment (assoomro) [reply src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs:80]: I find them helpful as they add more description, however most IDE's do that for you so there's really no need for them. I'll get rid of them.

File: src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs (hunk 3 above)
Comment (assoomro) [reply src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs:82]: 👍

File: src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs (hunk 4 above)
Comment (assoomro) [reply src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs:91]: 👍

File: src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs (hunk 6 above)
Comment (assoomro) [reply src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs:135]: 👍

File: src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs (hunk 5 above)
Comment (assoomro) [reply src/NHSD.BuyingCatalogue.Organisations.Api/Startup.cs:130]: 👍

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Healthcheck.3538.feature (hunk 7 above)
Comment (CallumBradburyBJSS) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Healthcheck.3538.feature:4]: all spaces, the mix of tabs and spaces keeps popping up

Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Healthcheck.3538.feature:4]: 👍

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Healthcheck.3538.feature (hunk 8 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Features/Organisations Healthcheck.3538.feature:16]: 👍

File: tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Utils/Settings.cs (hunk 9 above)
Comment (assoomro) [reply tests/NHSD.BuyingCatalogue.Identity.Api.IntegrationTests/Utils/Settings.cs:15]: 👍

File: src/NHSD.BuyingCatalogue.Identity.Common/Constants/HealthCheckTags.cs (hunk 1 above)
Comment (bjss-chris) [reply src/NHSD.BuyingCatalogue.Identity.Common/Constants/HealthCheckTags.cs:6]: Woops, Peter Griffin award to @bjss-chris!

You are of course correct so please ignore my previous comment.

**Pull request merged**